lib/graphql/execution/lookahead.rb

```diff
@@ -88,26 +91,24 @@ def selected?
       # Like {#selects?}, but can be used for chaining.
       # It returns a null object (check with {#selected?})
       # @return [GraphQL::Execution::Lookahead]
-      def selection(field_name, selected_type: @selected_type, arguments: nil)
-        next_field_name = normalize_name(field_name)
+      def selection(field_name, selected_type: @selected_type, arguments: nil, with_alias: false)
+        if !with_alias || selects?(field_name, arguments: arguments)
+          next_field_name = normalize_name(field_name)
+          next_field_defn = get_class_based_field(selected_type, next_field_name)
+          return lookahead_for_selection(next_field_name, next_field_defn, selected_type, arguments)
+        end
 
+        return alias_selections[field_name] if alias_selections.key?(field_name)
+
+        alias_node = lookup_alias_node(ast_nodes, field_name)
+        return NULL_LOOKAHEAD unless alias_node
+
+        next_field_name = alias_node.name
         next_field_defn = get_class_based_field(selected_type, next_field_name)
-        if next_field_defn
-          next_nodes = []
-          @ast_nodes.each do |ast_node|
-            ast_node.selections.each do |selection|
-              find_selected_nodes(selection, next_field_name, next_field_defn, arguments: arguments, matches: next_nodes)
-            end
-          end
 
-          if next_nodes.any?
-            Lookahead.new(query: @query, ast_nodes: next_nodes, field: next_field_defn, owner_type: selected_type)
-          else
-            NULL_LOOKAHEAD
-          end
-        else
-          NULL_LOOKAHEAD
-        end
+        arguments = @query.arguments_for(alias_node, next_field_defn)
+        arguments = arguments.is_a?(::GraphQL::Execution::Interpreter::Arguments) ? arguments.keyword_arguments : arguments
```

Comment:
It looks like, when `with_alias: true` is given, it ignores the provided `arguments` and instead, it returns _all_ matches. 

Was this change intentional? I _expect_ the default case (`arguments: nil`) to return all matches, but then any given arguments to be used as a filter for selected lookaheads. Why not leave that behavior in place? 

```suggestion
```

Reply:
In case when arguments are `nil` it will return _all_ fields that have the same type as the field with the provided alias, these two lines forced `#lookahead_for_selection` to find exactly one node with the corresponding alias. Without them `selection("f2", with_alias: true)` will return both `f1` and `f2` nodes, which looks confusing too 🙂

Sounds like it would be easier to keep `#selection` method as is and introduce a new one (`#alias_selection`), which will return a Lookahead with a single node matching the alias (because spec guarantees that there cannot be a second alias). I've refactored the code and specs. What do you think of this approach?

Reply:
Aaaaah, I think I finally understand what this is for. It seems like `arguments:` is being used to filter the selection set down to ones that _match the arguments of the aliased selection_. 

Unfortunately that doesn't cover all possibilities allowed by the spec. For example, this is perfectly valid: 

```graphql 
query {
  user(login: "dhh") { email } 
  dhh: user(login: "dhh") { bio }
}
```

What do you expect `lookahead.alias_selection("dhh").selections.map(&:name)` to return in that case? (I expect `bio`, is that right?) 

I pushed a test for this case in 84c1e135b, how do you think that should behave?